Add a rnode CLI command for checking whether given block has been finalized

itegulov opened the pull request:
## Overview
Title says it all


### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3806


### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>



### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

Commit:
Add isFinalized grpc service

Commit:
Add tests for last finalized API

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala:
```diff
@@ -0,0 +1,263 @@
+                   v3,
+                   bonds,
+                   Map(v1 -> genesis.blockHash, v2 -> genesis.blockHash, v3 -> genesis.blockHash)
+                 )
```
Why are you steering block creation by hand instead of just simulating real node behaviour by using 3 nodes? Those tests would be like 10x shorter with that approach.

See https://github.com/rchain/rchain/blob/dc1f0ef633618ff0671975ee88c6709e851286e6/casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala#L47-L77 for an example of creating a complex graph (yous seems simpler :))

itegulov replied:
Good point. I rewrote the PR tests by using three real nodes instead. Unfortunately, many older tests in `coop.rchain.casper.api` follow the "manual" block creation pattern.

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala:
```diff
@@ -0,0 +1,161 @@
+
+          b1 <- n1.addBlock(produceDeploys(0))
+          _  <- n2.receive()
+          _  <- n3.receive()
```
Thx for rephrasing the tests! :)

Could you add a `.publishBlock(deployDatums: DeployData*)(nodes: HashSetCasperTestNode: _*)` method to HashSetCasperTestNode, that would do `nodes.filter(_ != this).traverse_(_.receive())`? This way we could

`b1 <- n1.publishBlock((produceDeploys(0))(nodes)`, where `val nodes = List(n1, n2, n3)`

itegulov replied:
Done

KentShikama reviewed (approved):
Perhaps add a test for when it is the secondary parent of the finalized block?

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala:
```diff
@@ -0,0 +1,161 @@
+          isB2Finalized <- BlockAPI.isFinalized[Task](
+                            IsFinalizedQuery(hash = ProtoUtil.hashString(b2))
+                          )(Sync[Task], engineCell, SafetyOracle[Task], blockStore, logEff)
+          _ = isB2Finalized.right.value shouldBe IsFinalizedResponse(isFinalized = true)
```
The above 4 lines are duplicated 4 times in the added tests. I think they deserve a util method in that test class, which will also make the tests clearer (hopefully blockStore and logEff can be easily made implicit, so we'd only pass `engineCel` or `engine` to that method)

ArturGajowy commented on casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:
```diff
@@ -606,4 +606,45 @@ object BlockAPI {
+                               IsFinalizedResponse(isFinalized = true).asRight
+                             } else {
+                               IsFinalizedResponse(isFinalized = false).asRight
+                             }
```
This could just be `IsFinalizedResponse(isFinalized = lca == givenBlockMetadata).asRight`

itegulov commented:
~~While adding new tests as per @KentShikama's request, I realized that our current LCA implementation is buggy. I have ticketed my observations in https://rchain.atlassian.net/browse/RCHAIN-3828 and, since what is being added in this PR is higly dependent on LCA algorithm, I am temporarily switching my focus on RCHAIN-3828.~~

See https://github.com/rchain/rchain/pull/2719#issuecomment-529865613

itegulov commented:
See: https://github.com/rchain/rchain/pull/2719

Commit:
Rename LCA to lowestUniversalCommonAncestor to better reflect its nature

It is not the LCA in "classical" definition so it was decided to rename
this function in order to minimize confusion

itegulov commented:
@ArturGajowy PTAL. I have addressed your comments and also had to do a few more changes:
1. Renamed **current** lowestCommonAncestor function to lowestCommonForParentsAncestor
2. Implemented classical lowestCommonAncestor function which follows the classical LCA algorithm for DAGs
3. Used the newly introduced classical LCA for is-finalized call

Sorry for bringing a lot of confusion to this PR.

ArturGajowy commented on casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala:
```diff
@@ -91,6 +91,62 @@ object DagOperations {
+  /**
+    * LCA is the "lowest" (has highest blocknum) block which is an ancestor to both blocks.
+    */
+  def lowestCommonAncestorF[F[_]: Monad](
```
I think that rather than implementing this algorithm twice, we should express `isFinalized(given)` in terms of  
`.bfTraverse(lastFinalized)(n => n.parents.fliter(_.blockNum >= given.blockNum)).contains(given)`

Notice that the `.filter` above is just for early cutoff when `given` is above/sideways to `lastFinalized`

itegulov replied:
You are right. The initial plan was to reuse LCA but, since our current implementation is not what I thought it was, it is better to just BFS here.

ArturGajowy commented on casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala:
```diff
@@ -91,37 +91,80 @@ object DagOperations {
+    * We compute by finding the first block that is the "lowest" (has highest blocknum) block common
+    * for both blocks' parents.
+    */
+  def lowestCommonForParentsAncestorF[F[_]: Monad](
```
I think `for parents` is not precise here, as for the following:

```
A   B
 \ / \
  C   D
   \ / \ 
    E   F
     \ /
      G
```
LCFPA(A, B) == G, while - if we take parents of A and B, their LCA(A, B) == E

I think this should be called 'lowest universal common ancestor', which reflects the fact we require the ancestor to be the "Adam" for A and B, or "the nearest common ancestor of A, B, and all their ancestors"

itegulov replied:
Renamed it to lowestUniversalCommonAncestor and clarified the comment a bit further

ArturGajowy commented on casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala:
```diff
@@ -91,37 +91,80 @@ object DagOperations {
+  private val blockMetadataByNumDecreasing: Ordering[BlockMetadata] =
+    (l: BlockMetadata, r: BlockMetadata) => {
+      def compareByteString(l: ByteString, r: ByteString): Int =
+        l.hashCode().compareTo(r.hashCode())
```
I know you're not the author, but I find this a terrible hack:

1. given two byte strings, it's very unclear which one is going to be the 'smaller' one, as their hash is used for comparison,
2. this isn't always a tie breaker, as hashes can collide as well.

We should compare the byte strings lexicographically.

itegulov replied:
Replaced this with basically `Ordering.Iterable[Byte]`

ArturGajowy commented on casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala:
```diff
@@ -91,37 +91,80 @@ object DagOperations {
+
+      val ln = l.blockNum
+      val rn = r.blockNum
+      rn.compare(ln) match {
```
I've spent 15 minutes looking for why this thing sorts in descending order. A comment saying 'notice the inverted order of compared items, which makes the ordering descending' would hopefully help me.

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala:
```diff
@@ -0,0 +1,145 @@
+  val genesisParameters = buildGenesisParameters(
+    validatorKeyPairs,
+    Map(validatorPks(0) -> 10, validatorPks(1) -> 10, validatorPks(2) -> 5)
+  )
```
We can replace all the above setup with:
`  val genesisParameters = buildGenesisParameters(bondsFunction = _.zip(List(10L, 10L, 5L)).toMap)`

ArturGajowy replied:
Couldn't we somehow express the test with unitary weights? (all nodes have same weight?)

itegulov replied:
The weights are now unitary

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala:
```diff
@@ -0,0 +1,145 @@
+   */
+  "isFinalized" should "return true for ancestors of last finalized block" in effectTest {
+    HashSetCasperTestNode.networkEff(genesisContext, networkSize = 3).use {
+      case n1 +: n2 +: n3 +: _ =>
```
this can be

`case nodes @ n1 +: n2 +: n3 +: Seq()`

and then we don't need to declare `val nodes  = List(n1, n2, n3)` below

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala:
```diff
@@ -0,0 +1,145 @@
+  it should "return false for children and siblings of last finalized block" in effectTest {
+    HashSetCasperTestNode.networkEff(genesisContext, networkSize = 3).use {
+      case n1 +: n2 +: n3 +: _ =>
+        import n3.{blockStore, cliqueOracleEffect, logEff}
```
why do we differ the 'focused' node between those tests?

I have a hard time understanding which node sees what based on the picture. Maybe each node should have its own swim lane?

itegulov replied:
No particular reason. I changed it so that we always look from the n1 perspective.

Not sure what exactly do you mean by "swim lane" here but describing the whole state from all nodes POV would be overly verbose. Note that there is also "time" axis involved here: a node can get a block later than the other nodes (e.g. by missing it initially but fetching it as a dependency for a new one) which can affect node's decision on finalization. This is what happens in the second test (`it should "return false for children and siblings of last finalized block"`).

ArturGajowy replied:
I meant sth similar to how our dag pngs look like: each validator has a swim-lane, time is reflected by by the happens-before relationship encoded by justifications. No big deal :)

ArturGajowy reviewed (changes requested):
A couple comments :)

Commit:
Make LCUA compare ByteStrings by lexicographical byte ordering

Commit:
Add a comment about inverted order of compared arguments

ArturGajowy committed:
Add more test cases for lowestUniversalCommonAncestor

ArturGajowy reviewed (approved):
Good job! Thx for the fixups :)

I've added a commit with some more test cases for LUCA, I felt that test kinda looked through fingers on the fact that there's U in LUCA.

itegulov commented:
@ArturGajowy Thanks! I just wanted to say that I really appreciate your thoroughness and amount of attention you pay when it comes to reviewing PRs.

itegulov commented:
bors r+

ArturGajowy commented:
@itegulov ![oh stop it you](https://i.kym-cdn.com/entries/icons/original/000/010/809/ohstopitu.jpg)

Thx! :heart: This means a lot to me :)

itegulov commented:
bors r+
